### ISSUE TITLE
Error C4596: 'difference_type': illegal qualified name in member declaration when compiling range/concepts.hpp with MSVC141

### ISSUE THREAD

**Issue by ncook-vero** [boostorg/range#66]
Hi there,

I've been compiling boost 1.66.0 with MSVC 141 (Visual Studio 2017 15.5.6). 
When I do, I get Error C4596: 'difference_type': illegal qualified name in member declaration. This is caused by the namespace prefix at line 259 of concepts.hpp. (I'm assuming it's not a Visual Studio bug)
`             BOOST_DEDUCED_TYPENAME RandomAccessIteratorConcept::difference_type n;`

I've created a pull request, #65, with what I think is a fix, it works for me and passes CI.
Please let me know if the fix is suitable/appropriate. If it is, is there any chance it will make boost 1.67 ?

Thanks,

Nathan

**Comment by DanielaE** [comment]
@ncook-vero , @neilgroves ,
this change causes lots of compile failures in my runs of the test suite. For details see my comments to #65

**Comment by sav-ix** [comment]
Hello, everyone,

For builds using MSVC 2017 15.5.0 got errors:
```
compile-c-c++ bin.v2\libs\date_time\build\msvc-14.1\debug\address-model-64\threading-multi\gregorian\greg_month.obj

    call "bin.v2\standalone\msvc\msvc-14.1\address-model-64\architecture-x86\msvc-setup.bat"  >nul
 cl /Zm800 -nologo @"bin.v2\libs\date_time\build\msvc-14.1\debug\address-model-64\threading-multi\gregorian\greg_month.obj.rsp" 

greg_month.cpp
c:\libBOOST-1.68.0-dev\build\boost/range/concepts.hpp(257): error C3646: 'n': unknown override specifier
c:\libBOOST-1.68.0-dev\build\boost/range/concepts.hpp(264): note: see reference to class template instantiation 'boost::range_detail::RandomAccessIteratorConcept<Iterator>' being compiled
c:\libBOOST-1.68.0-dev\build\boost/range/concepts.hpp(257): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
...failed compile-c-c++ bin.v2\libs\date_time\build\msvc-14.1\debug\address-model-64\threading-multi\gregorian\greg_month.obj...
```
Reproduced after PR 795046f8fc30e359d0a4826cd5bb652d3ccf3855 and f1906e914eaaf76408a0a4e69c61c457c08abba8 were merged.

A workaround:
```
diff --git a/libs/range/include/boost/range/concepts.hpp b/libs/range/include/boost/range/concepts.hpp
index 6fef2ea..de36a22 100644
--- a/libs/range/include/boost/range/concepts.hpp
+++ b/libs/range/include/boost/range/concepts.hpp
@@ -253,7 +253,7 @@ namespace boost {
              }
          private:
  // MSVC 14.1 - avoid C4596: 'difference_type': illegal qualified name in member declaration
- #if defined(_MSC_VER) && _MSC_VER >= 1912 
+ #if defined(_MSC_FULL_VER) && _MSC_FULL_VER > 191225830
              BOOST_DEDUCED_TYPENAME difference_type n;
  #else
              BOOST_DEDUCED_TYPENAME RandomAccessIteratorConcept::difference_type n;
```
fixed it for me.
In case it can be used as a solution, a condition `_MSC_FULL_VER > 191225830` should be updated in order to correspond MSVC version, which introduced this issue.

**Comment by sav-ix** [comment]
Didn't check later MSVC versions yet. Assuming 15.5.6, mentioned in 1st post, should be fine.
@DanielaE, what makes you think, that this workaround won't be needed for future MSVC versions? (mean `191300000` and later) Or can you share link to related explanation?

**Comment by pdimov** [comment]
`typename difference_type n;` is ill-formed; it might be better to use something correct instead. Perhaps `typename RandomAccessIteratorConcept<Iterator>::difference_type n;` or `typename BidirectionalIteratorConcept<Iterator>::difference_type n;`

**Comment by jalegido** [comment]
According to Microsoft (https://msdn.microsoft.com/en-us/library/xwec471e.aspx) msvc creates and destroys global static objects in the same order they are in the code. In my experience, this means that it deppends of the order files in cl.exe.
For example, static objects in A.cpp are initialezed before than objects in B.cpp if you exec "cl A.cpp B.cpp". But if you change the order "cl B.cpp A.cpp", then the initialization order also reverse.

**Comment by ncook-vero** [comment]
I do like the idea of doing it correctly, apologies for my ill-formed solution - possibly it wasn't the solution I thought it was.

Of your suggestions, I prefer `typename RandomAccessIteratorConcept<Iterator>::difference_type n;` as it seems like the smallest change and doesn't introduce the question of what a `BidirectionalIteratorConcept` is to my mind - I wasn't trying to change or improve anything, I just wanted it to compile. 

Are you guys doing other work on this version of boost for this compiler version?
Is it worth doing, or am I distracting you from more useful things?

TL;DR: Should we just upgrade our boost to 1.67 (which says it's fine with VS2017) ?

**Comment by pabristow** [comment]
I have this problem on develop branch recently updated and current VS 15.7.2
      typename RandomAccessIteratorConcept<Iterator>::difference_type n;
and
     BOOST_DEDUCED_TYPENAME RandomAccessIteratorConcept<Iterator>::difference_type n;
both work for me, FWIW.

**Comment by jzmaddock** [comment]
Can someone please fix this?  It completely breaks Multiprecision apart from anything else.

**Comment by sav-ix** [comment]
+1 to something should be done at last.
After moving to MSVC 2017 15.7.3 got errors:
```
.\boost/range/concepts.hpp(257): error C3646: 'n': unknown override specifier
.\boost/range/concepts.hpp(264): note: see reference to class template instantiation 'boost::range_detail::RandomAccessIteratorConcept<Iterator>' being compiled
.\boost/range/concepts.hpp(257): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
```
A possible solution:
- revert https://github.com/boostorg/range/pull/65/commits/f1906e914eaaf76408a0a4e69c61c457c08abba8 and https://github.com/boostorg/range/pull/65/commits/795046f8fc30e359d0a4826cd5bb652d3ccf3855 (this fixes builds using MSVC 2017 15.7.x and do not break builds using mingw-w64 8.1.0 and Windows ICC 2018 Update 3),
- close this issue as out of scope (it was a compiler bug in MSVC 2017 15.5.x and should not affect Boost code),
- advice to all Boost Users, who experience this error, to upgrade their build tools.

How's that, @ncook-vero ?

**Comment by ncook-vero** [comment]
Hi @sav-ix, we (Vero) have now upgraded to VS 2017 15.7.4, so we should be fine with your suggestion if we can build boost with it in it's unadultered-by-myself form - we prefer that really..

Essentially, I think @DanielaE hit the nail on the head: "Therefore, it's application must be restricted to this short range". You could either fix what appears to be an edge case, or advise upgrading build tools.

I know what I'd do - advise upgrade and keep clean code :)

Let me build a vanilla boost 1.67 with VS 15.7.4, and I shall get back to you ASAP

**Comment by pabristow** [comment]
I'm confused.  Is Peter Dimov wrong when he says the original was ill-formed?  Sounds unlikely but this is above my paygrade?
I'd like the current VS at the time of release of 1.68 to work with the current VS version at that release time (and to work with the then current develop branch too of course).  It's a show stopper for Multiprecision (and other ?)

**Comment by ncook-vero** [comment]
@pdimov @sav-ix  - I still experience this with VS 2017 15.7.4.
Please see screenshot below demonstrating this. Please excuse my poor redactions of our closed source.

![image](https://user-images.githubusercontent.com/36889813/42449944-72e47884-837a-11e8-890a-8f7d7fa3523b.png)

However, my ill-formed fix continues to be a fix for this, for us at least, it would appear.

**Comment by pdimov** [comment]
Do you have a small program I can try?

**Comment by ncook-vero** [comment]
It's not very small I'm afraid, and it's rather closed source. I'll see if I can put together a reproduction. outside of our product.
I'm also juggling a V8 upgrade and various rebases today. I will get back to you A.S.A.P

**Comment by jzmaddock** [comment]
>However, my ill-formed fix continues to be a fix for this, for us at least, it would appear.

Well this is very strange, because for me with a fresh install of 15.7.4 and a program that consist of just:

```
#include <boost/range/concepts.hpp>
```

I get the:

```
1>c:\data\boost\boost\boost\range\concepts.hpp(257): error C3646: 'n': unknown override specifier
```

error.

This is debug, x64, no pre-compiled headers.

**Comment by pdimov** [comment]
Interesting. That `#include` compiles for me both with and without `/permissive-`.

**Comment by pdimov** [comment]
OK, @ncook-vero, does https://github.com/boostorg/range/pull/70 work for you?

**Comment by ncook-vero** [comment]
Ok, here's what I've been trying - building a boost 1.67 containing one of these `difference_type` definitions, then building our project against it. Here's my sketchpad:

    //#if defined(_MSC_VER) && _MSC_VER >= 1913
    //		BOOST_DEDUCED_TYPENAME difference_type n;
    //#else
    //		BOOST_DEDUCED_TYPENAME RandomAccessIteratorConcept::difference_type n;
		        BOOST_DEDUCED_TYPENAME BidirectionalIteratorConcept<Iterator>::difference_type n;
    //#endif

#70 didn't work for me I'm afraid:
`BOOST_DEDUCED_TYPENAME RandomAccessIteratorConcept::difference_type n` yields `Error	C4596	'difference_type': illegal qualified name in member declaration (compiling source file xxxxxx.cpp)	xxxxxxx d:\development\xxxxxxx\3rdparty\boost\boost\range\concepts.hpp	258`

However, based on a previous suggestion, I tried
`BOOST_DEDUCED_TYPENAME BidirectionalIteratorConcept<Iterator>::difference_type n;` and it worked. Given that this is essentially asking the parent class and template type to provide the definition, I think  this essentially demonstrates the (fluke) goodness of my original fix (I think it is essentially asking the compiler the same thing, isn't it?). 

Am I missing something?

RE: /permissive-, we compile with /WAll, warnings-as-errors, etc. I saw /Zc:rvalueCast mentioned in the /permissive- MSDN page as turned on by /permissive-  .. I also saw /Zc:rvalueCast in our project settings, so we are compiling with /permissive-

**Comment by pdimov** [comment]
OK, I'll do another PR with `BidirectionalIterator` instead.

**Comment by pdimov** [comment]
No success in reproducing the problem in a self-contained test?

**Comment by ncook-vero** [comment]
I'm attempting to create one now similar to @jzmaddock, just need to rebuild boost

**Comment by ncook-vero** [comment]
Reporting back on a repro outside of our product..

VS2017 15.7.4, Boost 1.67

In a harness (new Win32 console app), it is my change in concepts.hpp that causes the `error C3646: 'n' : unknown override specifier` - so I think I have reproduced what everyone else is seeing. The harness was compiled with /WAll /WX /permissive- /Yu in debug and /W4 /WX /permissive- /Yu in release (x64 only). 

Worth noting that using `BOOST_DEDUCED_TYPENAME RandomAccessIteratorConcept::difference_type n;` in place of my change (i.e., the original code) fixes the `error C3646: 'n' : unknown override specifier` in the harness.

However, returning to our product which includes the very same header, with the same original-code-not-my-fix experimental change from the harness, the result is somewhat reversed and I get `error C4596: 'difference_type': illegal qualified name in member declaration (compiling source file WeaveUtils.cpp)`

**Comment by vinniefalco** [comment]
This is still a problem in 1.68 beta 1. Attempting to build all the boost libraries on my machine with /permissive- produces this output https://gist.github.com/vinniefalco/40b2e8e6a283377330258ce1e3281b6c

**Comment by Flamefire** [comment]
@vinniefalco Would you mind testing if removing the `BOOST_DEDUCED_TYPENAME` from the line at `C:\Users\vinnie\src\boost\boost/range/concepts.hpp(257)` (and only there) solves this for you?

**Comment by vinniefalco** [comment]
Removing BOOST_DEDUCED_TYPENAME still produces an error. However, changing it as follows resolves the error (I am on the latest VS)
```
#if defined(_MSC_VER) && _MSC_VER >= 1912 
             BOOST_DEDUCED_TYPENAME BidirectionalIteratorConcept<Iterator>::difference_type amount;
 #else
             BOOST_DEDUCED_TYPENAME RandomAccessIteratorConcept::difference_type amount;
 #endif
```

**Comment by Flamefire** [comment]
Does it produce the same error or a different one? And lastly: Would you mind testing `BOOST_DEDUCED_TYPENAME RandomAccessIteratorConcept::difference_type amount;` too (or remove the first part of the `#if`)? If that works than previous discussion is right: The work-around should only be for a specific MSVC version or completely removed as the non-MSVC code is actually correct.

**Comment by vinniefalco** [comment]
* Same error

* `BOOST_DEDUCED_TYPENAME RandomAccessIteratorConcept::difference_type n;` also compiles correctly

**Comment by vinniefalco** [comment]
I suggest the following as a fix:
```
 #if defined(_MSC_VER) && _MSC_VER < 1912 
    BOOST_DEDUCED_TYPENAME difference_type n;
 #else
    BOOST_DEDUCED_TYPENAME RandomAccessIteratorConcept::difference_type n;
 #endif
```

**Comment by vinniefalco** [comment]
All the boost libraries build correctly with `BOOST_DEDUCED_TYPENAME RandomAccessIteratorConcept::difference_type n`

**Comment by Flamefire** [comment]
Just seen that develop already contains the fixes: #70 and #71
This does not require macros. To summarize everything said and confirmed so far it is a bug in `_MSC_VER==1912` only (original fix said >= 1912, but others mentioned 1913 is not affected and a more fine grained `#if defined(_MSC_FULL_VER) && _MSC_FULL_VER > 191225830 && _MSC_FULL_VER < 191300000` was suggested)

Please get #70 and #71 merged into master for the 1.68 version and this can be closed.

**Comment by ncook-vero** [comment]
Okay (seems likely that I would stumble into a Visual Studio bug, that is my kind of luck), but what about our failed build? Simply removing `/permissive-` made it fail at that location - granted that the example code is mad..

**Comment by Flamefire** [comment]
Did you test with those 2 PRs? Or simply replace the full `#if` stuff here with `BOOST_DEDUCED_TYPENAME BidirectionalIteratorConcept<Iterator>::difference_type n;` and test that. Are there still errors in your build then? If so, could you post that error? Would be very interesting!

**Comment by pdimov** [comment]
You wrote above

> However, based on a previous suggestion, I tried BOOST_DEDUCED_TYPENAME BidirectionalIteratorConcept<Iterator>::difference_type n; and it worked.

Is this not true?

**Comment by ncook-vero** [comment]
It is true.  At least I think it is. I suggest trying my ` TestBoostConcepts.cpp` sample in a VS 2017 15.7.4, toggling /permissive- in project settings, and trying the various difference_type definitions in concepts.hpp, that's what I did. /permissive- seemed to be key to the issue. Some of our projects use /permissive- and some don't. We need boost to work with all of them. The TestBoostConcepts.cpp code comes from a section of code in a project that is not compiled with /permissive-.

**Comment by Flamefire** [comment]
So where is the problem? If `BOOST_DEDUCED_TYPENAME BidirectionalIteratorConcept<Iterator>::difference_type n;` works for you then you did not test what I sugested in https://github.com/boostorg/range/issues/66#issuecomment-404427400

**Comment by pdimov** [comment]
Unfortunately, your sample compiles for me no matter which definition of `n` I use in `concepts.hpp`. My command line does not match exactly:

```
/GS /GL /W4 /Gy /Zc:wchar_t /I"/boost-git/develop" /Zi /Gm- /Ox /Ob2 /Fd"x64\Release\vc141.pdb" /Zc:inline /fp:precise /D "NDEBUG" /D "_CONSOLE" /D "_UNICODE" /D "UNICODE" /errorReport:prompt /WX- /Zc:forScope /Gd /Oi /MD /FC /Fa"x64\Release\" /EHsc /nologo /Fo"x64\Release\" /Fp"x64\Release\testbed2017.pch" /diagnostics:classic 
```

but I don't see how any of those options would affect the front end. What do you have in `stdafx.h`?

**Comment by pdimov** [comment]
In either case, the develop branch now contains the `BidirectionalIteratorConcept` definition, so if you could check that your product compiles with it, we'd be done.

**Comment by ncook-vero** [comment]
`stdafx.h` contains:

    // stdafx.h : include file for standard system include files,
    // or project specific include files that are used frequently, but
    // are changed infrequently
    //
    
    #pragma once
    
    #define _CRT_SECURE_NO_WARNINGS
    #define _SCL_SECURE_NO_WARNINGS
    
    #include "targetver.h"
    
    // TODO: reference additional headers your program requires here

I'll check again, bear with me. We are using /WAll /WX, so we must disable many warnings in the product, as in the sample. We disable others in the product, I just disabled what I had to so that the example would 'compile' at least as far as concepts.hpp.

**Comment by pdimov** [comment]
Why didn't you use the current contents of the develop branch?

https://github.com/boostorg/range/blob/70d1727ed301f2852ba911aacdcbcc57bb0ac92d/include/boost/range/concepts.hpp#L254-L257

**Comment by ncook-vero** [comment]
I'm `#if defined(_MSC_VER) && _MSC_VER >= 1913`

**Comment by Flamefire** [comment]
Can this be simply closed please? The bug is obviously fixed/handled in develop and working for everyone so far.

**Comment by pdimov** [comment]
OK, I've taken the liberty of merging the fix to master.

**Comment by Flamefire** [comment]
Great! Make sure it gets included in the 1.68 release, please.

**Comment by neilgroves** [comment]
Peter,

Thank you enormously for your massive help in this. I was massively time
constrained. Your help is really appreciated.

Thanks,
Neil

**Comment by pabristow** [comment]
As well as thanks to/from everyone, would it be useful to record why this concept fix, especially why bidirectional rather than randomaccess?

Paul

**Comment by Flamefire** [comment]
> As well as thanks to/from everyone, would it be useful to record why this concept fix, especially why bidirectional rather than randomaccess?

The type is not defined in the current type but in the parent type like here:

```
template<class T>
struct Foo{
  using type = T;
};
template<class T>
struct Bar: public Foo<T>{
  static typename Foo<T>::type value = T(0);
};
```

Not sure if `static typename Bar::type value = T(0);` is covered by the standard, but at least MSVC chocked on that (in at least 1 version) and I remember similar problems in other libraries too (using the parent type instead of the current fixed a problem)
